File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Maximum_Twin_Sum_of_a_Linked_List/sunghyun.cpp
#include <climits>
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        int lllen = 0;
        ListNode* temp = head;

        while(temp != nullptr){
            lllen++;
            temp = temp->next;
        }

        int mid = lllen / 2;

        ListNode* firstptr = head;
        int midval = 0;
        ListNode* secondptr = head;
        int maxsum = INT_MIN;

        while(midval != mid){
            secondptr = secondptr->next;
            midval++;
        }

        ListNode* prev = nullptr;
        ListNode* currptr = secondptr;

        while(true){
            ListNode* nextptr = currptr->next;
            currptr->next = prev;
            prev = currptr;
            if(nextptr == nullptr){
                break;
            }
            currptr = nextptr;
        }

        ListNode* reverseptr = currptr;

        while(reverseptr != nullptr){
            int pairsum = firstptr->val + reverseptr->val;
            if(pairsum > maxsum){
                maxsum = pairsum;
            }
            firstptr = firstptr -> next;
            reverseptr = reverseptr -> next;
        }

        return maxsum;
    }
};

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Maximum_Twin_Sum_of_a_Linked_List/sunghyun.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def pairSum(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: int
        """
        lllen = 0
        temp = head

        while(temp != None):
            lllen = lllen + 1
            temp = temp.next
        
        mid = lllen / 2

        firstptr = head
        midval = 0
        secondptr = head
        maxsum = float('-inf')

        while(midval != mid):
            secondptr = secondptr.next
            midval = midval + 1
        
        prev = None
        currptr = secondptr

        while(True):
            nextptr = currptr.next
            currptr.next = prev
            prev = currptr
            if (nextptr == None):
                break
            currptr = nextptr
        
        reverseptr = currptr

        while(reverseptr != None):
            pairsum = firstptr.val + reverseptr.val
            if (pairsum > maxsum):
                maxsum = pairsum
            firstptr = firstptr.next
            reverseptr = reverseptr.next
        
        return maxsum
            
        